src/intersections/boxes.jl

```diff
@@ -34,3 +34,32 @@ function intersecttype(f::Function, b1::Box{Dim,T}, b2::Box{Dim,T}) where {Dim,T
     return FaceTouchingBoxes(Box(u, v)) |> f
   end
 end
+
+"""
+    intersecttype(function, ray, box)
+
+Calculate the intersection type of a `ray`` and a `box`` and apply function `f` to it.
+
+## References
+
+* Williams A, Barrus S, Morley R K, et al., 2005. [An efficient and robust ray-box
+  intersection algorithm]
+  (https://dl.acm.org/doi/abs/10.1145/1198555.1198748)
+"""
+function intersecttype(f::Function, r::Ray{3,T}, b::Box{3,T}) where {T}
+  sign = (x -> x < 0 ? 2 : 1).(direction(r))
+  invdir = one(T) ./ direction(r)
+  bounds = (b.min, b.max)
+
+  tmin = maximum((bounds[sign[1]] - origin(r)) .* invdir)
+  tmin = max(tmin, zero(T))
+  tmax = minimum((bounds[3 - sign[1]] - origin(r)) .* invdir)
```

Comment:
Can you please add a comment explaining the idea behind these lines? They look a bit more complicated than they coul be I guess. If you can describe in words we can think in a better code to describe the idea.

Reply:
@juliohm I guess you are confused on `bounds[sign]` and `bounds[3 .- sign]`. Someone told me this magic can get 1.5x faster than use comparation.

The original version looks like

```julia
if (invdir[1] >= 0)
  tmin = (max[1] - origin(r)[1]) * invdir[1]
  tmax = (max[1] - origin(r)[1]) * invdir[1]
else
  tmax = (max[1] - origin(r)[1]) * invdir[1]
  tmin = (max[1] - origin(r)[1]) * invdir[1]
end
```

Use the array access to avoid this comparation. This trick may hard to added in the code via comments.

Besides, I found I wrote a wrong code previously. I am trying to fix this.